source/h1_encoder.c

```diff
@@ -319,6 +351,404 @@ int aws_h1_encoder_start_message(
+            break;
+    };
+    return AWS_OP_SUCCESS;
+}
+
+void aws_h1_lock_chunked_list(struct aws_http1_chunks *body_chunks) {
+    AWS_PRECONDITION(body_chunks);
+    int err = aws_mutex_lock(&body_chunks->lock);
+    AWS_ASSERT(!err);
+    (void)err;
+}
+
+void aws_h1_unlock_chunked_list(struct aws_http1_chunks *body_chunks) {
+    AWS_PRECONDITION(body_chunks);
+    int err = aws_mutex_unlock(&body_chunks->lock);
+    AWS_ASSERT(!err);
+    (void)err;
+}
+
+bool aws_h1_get_next_stream_chunk(struct aws_http1_chunks *body_chunks, struct aws_http1_stream_chunk **chunk_out) {
+    AWS_PRECONDITION(body_chunks);
+    AWS_PRECONDITION(chunk_out);
+    bool has_next_chunk = true;
+    /* Begin critical section */
+    aws_h1_lock_chunked_list(body_chunks);
+    if (!aws_linked_list_is_valid(&body_chunks->chunk_list)) {
+        has_next_chunk = false;
+        goto release_lock;
+    }
+    if (aws_linked_list_empty(&body_chunks->chunk_list)) {
+        *chunk_out = NULL;
+        body_chunks->paused = true;
+        has_next_chunk = false;
+    } else {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&body_chunks->chunk_list);
+        *chunk_out = AWS_CONTAINER_OF(node, struct aws_http1_stream_chunk, node);
+        aws_linked_list_node_reset(node);
+        AWS_POSTCONDITION(chunk_out);
+    }
+release_lock:
+    aws_h1_unlock_chunked_list(body_chunks);
+    /* End critical section */
+    return has_next_chunk;
+}
+
+static size_t s_calculate_chunk_line_size(struct aws_http1_chunk_options *options) {
+    size_t chunk_line_size = MAX_ASCII_HEX_CHUNK_STR_SIZE + CRLF_SIZE;
+    for (size_t i = 0; i < options->num_extensions; ++i) {
+        struct aws_http1_chunk_extension *chunk_extension = options->extensions + i;
+        chunk_line_size += sizeof(';');
+        chunk_line_size += chunk_extension->key.len;
+        chunk_line_size += sizeof('=');
+        chunk_line_size += chunk_extension->value.len;
+    }
+    return chunk_line_size;
+}
+
+static bool s_populate_chunk_line_buffer(struct aws_byte_buf *chunk_line, struct aws_http1_chunk_options *options) {
+    bool wrote_chunk_line = true;
+    wrote_chunk_line &= aws_write_chunk_size(chunk_line, options->chunk_data_size);
+    for (size_t i = 0; i < options->num_extensions; ++i) {
+        wrote_chunk_line &= aws_write_chunk_extension(chunk_line, options->extensions + i);
+    }
+    wrote_chunk_line &= aws_write_crlf(chunk_line);
+    return wrote_chunk_line;
+}
+
+int aws_chunk_line_from_options(struct aws_http1_chunk_options *options, struct aws_byte_buf *chunk_line) {
+    size_t chunk_line_size = s_calculate_chunk_line_size(options);
+    if (AWS_OP_SUCCESS != aws_byte_buf_init(chunk_line, options->chunk_data->allocator, chunk_line_size)) {
+        return AWS_OP_ERR;
+    }
+    if (!s_populate_chunk_line_buffer(chunk_line, options)) {
+        /* The caller will free the chunk, which will free the chunk line in the process */
+        return AWS_OP_ERR;
```

Comment:
returning AWS_OP_ERR without any aws_raise_error() in the callstack.
Seems like this can only go wrong if s_calculate_chunk_line_size() miscalculated. If you are extremely confident in that calculation, you could simply AWS_ASSERT() that it wrote all. If you think it might have edge-cases you could log the error and `return aws_raise_error(AWS_ERROR_INVALID_STATE)`

Reply:
I'll move it to an assert. The size is calculated internally and the buffer is allocated off of it, so it would be a programming error to miscalculate the actual written size.